Resolve Issue#118

olyson opened the pull request:
### Description of changes
Resolve [issue#118](https://github.com/ESCOMP/CTSM/issues/118)
### Specific notes
I've renamed the variables as discussed in [issue#118](https://github.com/ESCOMP/CTSM/issues/118) throughout the code.
I've also made a couple of minor changes to fix a couple of potential problems with these variables as described in the branch commit logs.
I tested for bfb before changing the history field variables themselves. These changes are all bfb (with the exception of QSNOEVAP) for a 1 month global 2deg simulation, but may not be bfb under all conditions.  QSNOEVAP is not bfb because I've added in qflx_ev_snow for lakes.

Contributors other than yourself, if any:
Variable names discussed with Bill Sacks, variable functions discussed with Sean Swenson.

CTSM Issues Fixed (include github issue #):
[118](https://github.com/ESCOMP/CTSM/issues/118)

Are answers expected to change (and if so in what way)?
See specific notes above.

Any User Interface Changes (namelist or namelist defaults changes)?
No.

Testing performed, if any:
See specific notes above.

billsacks commented:
Thanks a lot, @olyson . I'll take a look at this.

One initial question: Did you address this comment in this PR?

> However, it seems like the qflx_ev_snow variable on the history file (QSNOEVAP) will have evaporation from urban classified as snow evaporation which will be wrong when there is no snow. And possibly other variables derived from qflx_ev_snow.

I'm a bit lost as to what did and didn't get addressed from #118 , and just wanted to check if the above comment is addressed sufficiently, still needs to be addressed, or if you decided that nothing should be done about that.

olyson commented:
I decided that nothing needed to be done about that comment since urban eflx_ev_snow is essentially being treated the same as for soil, crop, etc., in that eflx_ev_snow will represent evaporation from soil when there is no snow.  This is also consistent with how I've added eflx_ev_snow for lake.
In the end, QSNOEVAP (eflx_ev_snow) represents evaporation from snow only if there is at least one snow layer.  This is consistent among all the landunits now.

billsacks commented:
Sounds good - thanks.

olyson commented:
As Bill noted in his review, qflx_liqdew_ and qflx_soliddew_ were reversed throughout the code.  I've fixed this on the branch.  These changes are still bfb with the exception of QSNOEVAP as before.

olyson commented on src/biogeophys/WaterFluxBulkType.F90:
```diff
@@ -205,6 +203,16 @@ subroutine InitBulkHistory(this, bounds)
          ptr_patch=this%qflx_snotempunload_patch, set_lake=0._r8, c2l_scale_type='urbanf')
 
+    ! QSNOEVAP is evaporation from snow but only when snow is present (snl<0), otherwise, it is 
+    ! equivalent to qflx_ev_soil   
```
I think that's a good idea.  I've tried to shorten it without being too cryptic.  Line length is 117 characters but that is below the max of 132.  I don't think you can use '&'  in the middle of quotes  right?  I've updated my branch.

billsacks replied:
Looks good to me now - thanks. I think this is a nice, concise description. For the future, though, you can make longer strings by either:

(1) Using a continuation character at the end of the first line and start of the next:

```Fortran
  this = 'is a string with a con&
       &tinuation character'
```

(2) Concatenating two strings:

```Fortran
  this = 'is a string made from' // &
       ' two concatenated strings'
```

But again, I'm happy with what you have here, so I'l start testing with this.

(There is likely also a hard-coded limit on the length of the long_name field itself.)

billsacks commented:
@olyson - I started testing on this, but many of the tests are failing. I spot-checked two (`SMS_D.f10_f10_musgs.I2000Clm50BgcCrop.izumi_intel.clm-crop` on izumi and `ERP_D_Ld5.f10_f10_musgs.I2000Clm50Sp.cheyenne_intel.clm-reduceOutput` on cheyenne), and they are both failing with a floating point exception on line 273 of SoilFluxesMod.F90:

```Fortran
            qflx_ev_snow(p) = qflx_ev_snow(p) * egirat(c)
```

I wonder if `qflx_ev_snow` might be starting out uninitialized in some cases with your new code?

This also makes me think that I should "teach you to fish", if I haven't already, by showing you the new method for running the test suite. It's a lot easier than it used to be. So once you fix this problem, I'd be happy to show you the process so that you can do further testing of this and future tags yourself, if you're open to it.

olyson commented:
I'm happy to learn the new test suite method.  You shouldn't be spending any time troubleshooting this.

olyson commented:
If the system testing guide is up to date, I can try following that first.

https://github.com/ESCOMP/CTSM/wiki/System-Testing-Guide

olyson commented:
I fixed the problem and ran those two tests.  The one on cheyenne passed, although a comparison to baseline couldn't be made because the test didn't seem to exist for ctsm1.0.d081?  The test on izumi passed, except for the baseline comparison, which failed only because of the QSNOEVAP variable (expected).
I moved on to the aux_clm test suite on izumi.  Everything passed except for most of the baseline comparisons, which all failed only because of the QSNOEVAP variable (I grepped all of the cprnc.out files for RMS).
So I'll move on to the aux_clm tests on cheyenne.

billsacks commented:
That's great to hear. If you haven't already, you should hold off on running the full test suite on cheyenne, because @negin513 is in the process of making a tag.

olyson commented:
Will do.

olyson commented:
Test results on cheyenne:
6 tests failed because QFLX_SUB_SNOW_ICE is a requested history variable and it doesn't exist anymore
2 expected fails
9 TPUTCOMP fails
3 MEMCOMP fails
1 unexplained failure (fails with MPT: Received signal 15 in cesm.log, and 2020-01-30 14:50:39: ERROR: Job killed due to receiving signal 15 (SIGTERM) in TestStatus.log:
SMS_D_Vnuopc.f10_f10_musgs.I2000Clm50SpRsGs.cheyenne_intel.clm-default

Most baselines comparisons fail because of QSNOEVAP, which is expected.

I assume the TPUTCOMP and MEMCOMP failures are due to system variability.

Should I remove QFLX_SUB_SNOW_ICE from those 6 tests and rerun them?

Any ideas on the unexplained failure?  I can't seem to find any other useful error messages.

I tried to rerun testing on izumi, even though the tests passed previously, in order to generate baselines, but I now get an error on all the builds (except gnu).  Seems like at least one of the compilers is a license issue.

olyson commented:
Ok, looks like the unexplained failure was due to a wallclock issue.  The default setting in env_workflow.xml is 0:30, which I think is being interpreted as 30 seconds.  The test seems to need more than that.  I set it to 1:00, which I think is being interpreted as 1 minute, and the test passes.  The baseline comparison fails because: ctsm1.0.dev081: ERROR BFAIL some baseline files were missing

negin513 commented:
@olyson 
Correct! `SMS_D_Vnuopc.f10_f10_musgs.I2000Clm50SpRsGs.cheyenne_intel.clm-default` fails because in `cime_config/testdefs/testlist_clm.xml` the default wallclock time is set as `0:30` instead of `0:30:00`. I fixed this issue in my PR so hopefully we won't see this in future.

billsacks commented:
@olyson thanks for the detailed error report.

Regarding the `QFLX_SUB_SNOW_ICE` failure: Please remove or change the reference to this history variable in `cime_config/usermods_dirs/_includes/output_base/user_nl_clm` then you can rerun just the failing tests. When rerunning these tests, I would do it as follows (it won't work to simply rebuild & resubmit these, because the issue is with the user_nl file that is copied in at create_newcase time, so to be safe you'll need to completely recreate these tests):

(1) Manually remove the baseline directories for just these tests from wherever you generated baselines

(2) Create a text file containing just these 6 tests; I'll assume it's called 'tests_to_rerun'

(3) Run the `run_sys_tests` command, but instead of `-s aux_clm` give it the argument `-f /path/to/tests_to_rerun`; also give it something like `--walltime 2:00:00` (for this mode of operation, all tests have the same max wallclock time).

I'm going to check in with Mark Moore about the izumi issues. Other than that, this seems good. (I can write a ChangeLog entry based on what you wrote at the top of this PR.)

olyson commented:
I've replaced those old variable names with the new ones in usermods_dirs and repeated those six tests.  They all passed, except for baseline comparisons (due to QSNOEVAP as expected) and NLCOMP (expected).
I've committed the changes to my branch.

billsacks commented:
I have updated this branch to point to a cime tag that fixes intel and cprnc issues on izumi, and have run the test suite on izumi. All tests pass, with expected baseline failures in QSNOEVAP but nothing else. I'll bring this to master shortly.
